### DISCUSSION TITLE
what's suggested way of implementing dragging in v7

### DISCUSSION THREAD

**Discussion by tiye** [pixijs/pixijs#8813] [General]
updates: I moved to new events `globalpointermove` introduced later and it worked for my case https://github.com/pixijs/pixijs/releases/tag/v7.2.0 .

-----

Context is v7 removed InteractionManager and my implementation of dragging no longer worked. My [original code](https://github.com/Phlox-GL/phlox/blob/main/compact.cirru#L1054-L1087) was in Calcit which compiles to JavaScript and it roughly looks like:

```cirru
let
   ;; "the position passed to component"
    position props.position
  container
    {} $ :position position
    circle $ {} (:radius radius)
      :position $ [] 0 0
      :fill fill
      :alpha alpha
      :on $ {}
        :pointerdown $ fn (e d!)
          ;; "mousedown happen, track original click position, track original position value"
          let
              x e.data.global.x
              y e.data.global.y
            setState $ merge state
              {}
                :dragging? true
                :x0 $ [] x y
                :p0 position
        :pointermove $ fn (e d!)
          when (:dragging? state)
            ;; "during movement, compare new position with mousedown positions, call onChange to set new position"
            let
                x e.data.global.x
                y e.data.global.y
              let
                  x0 state.x0
                on-change
                  complex/add state.p0
                    []
                      * unit $ - x (first x0)
                      * unit $ - y (last x0)
                  , d!
        :pointerup $ fn (e d!)
          ;; "release dragging state"
          setState $ assoc state :dragging? false
        :pointerupoutside $ fn (e d!)
          setState $ assoc state :dragging? false
```

since `pointermove` is trigger easier even mouse/pointer moves out of my `circle` that responds for dragging, dragging is quite easily implemented.

In EventSystem in v7, `pointermove` is no longer triggered outside `circle`. What's the suggested way of listening to the events outside the shape?

I could think of a way listening to global `mousemove` event and build a dragging like this, but it would probably mess up with the declarative code since event no longer dispatched the shape.(didn't look into implementation in v7 yet).

**Comment by bigtimebuddy** [comment]
Interaction events in PixiJS now behave like the DOM in v7. This was intentional to align around behavior that would be familiar with developers, but obviously impacts the behavior with pointermove/mousemove/touchmove. 

Like the DOM, *move events are now local. This means that if you are outside the bounds of the object, you will _not_ receive a move event. Generally, you should consider adding move events to the stage/parent _instead_ of the object itself. 

Here is an example using v7:
https://jsfiddle.net/bigtimebuddy/spnv4wm6/

**Reply by tiye** [reply]
while it could work, does PIXI shapes support dispatching custom events in the new Event system?

I'm thinking of dispatching my own `move-in-stage` event so that my previous logics could be reused.

**Reply by bigtimebuddy** [reply]
Yes, you can add your own events. All displayObjects extend event emitter so you can emit that event. Good solution!

**Reply by tiye** [reply]
~~and now problem turns into "how do I find the shape that want to listen to global mouse moving events"? ....while `pointerdown` happened.~~

let's hope that `event.target` is always non `null`.

**Reply by tiye** [reply]
well
```
Error: DisplayObject cannot propagate events outside of the Federated Events API
```
only old way to try tomorrow...

**Reply by tiye** [reply]
events happen in a limited area,, should be enough for now..
```js
// The stage will handle the move events
app.stage.interactive = true;
app.stage.hitArea = app.screen;
```
but also dispatching the sam event caused `Uncaught RangeError: Maximum call stack size exceeded`, how can I rename the event when it does not have a setter on `name` field...

**Reply by gonulbuya** [reply]
app.stage.interactive = true;
app.stage.hitArea = app.screen;
arabaClastan2.on('pointerdown', onDragStart);
app.stage.on('pointerup', onDragEnd);
app.stage.on('pointerupoutside', onDragEnd);



function onDragStart(e) {


    this.alpha = 0.5;
    dragTarget = this;

    dragTarget.offset = {x: dragTarget.x - dragTarget.parent.toLocal(e.global).x, y: this.y - dragTarget.parent.toLocal(e.global).y};
    app.stage.on('pointermove', onDragMove);
}

function onDragMove(e) {
    
    var _this = e.currentTarget;
    
    if (dragTarget) {
        // const newPosition = _this.toLocal(event.global);
        // _this.position.x = newPosition.x+_this.offset.x;
        // _this.position.y = newPosition.y+_this.offset.y;

        //dragTarget.parent.toLocal(event.global)


         const newPosition = dragTarget.parent.toLocal(e.global)
         dragTarget.x = newPosition.x+dragTarget.offset.x;
         dragTarget.y = newPosition.y+dragTarget.offset.y;

    }

    
}

function onDragEnd(e) {

    if (dragTarget) {
        app.stage.off('pointermove', onDragMove);
        dragTarget.alpha = 1;
        dragTarget = null;
    }

    
}

**Reply by gonulbuya** [reply]
> PixiJS'deki etkileşim olayları artık v7'deki DOM gibi davranıyor. Bu, geliştiricilerin aşina olduğu, ancak pointermove/mousemove/touchmove ile davranışı açıkça etkileyen davranışlar etrafında hizalanmak için kasıtlıydı.
> 
> DOM gibi, *move olayları da artık yereldir. Bu, nesnenin sınırları dışındaysanız bir taşıma olayı almayacağınız _anlamına gelir. _Genel olarak, nesnenin kendisi yerine _sahneye_ /üst öğeye taşıma olayları eklemeyi düşünmelisiniz .
> 
> İşte v7 kullanan bir örnek: https://jsfiddle.net/bigtimebuddy/spnv4wm6/

This code of yours and the one I wrote a while ago are doing the same task, right? If what you wrote is better, I would like to use it.

**Comment by jakerdy** [comment]
Local `XXXmove` events are understandable, but why not add something like `mousemoveoutside` or `mousedrag` or  base class poroperty `allowOutsideMove` on EventTarget. There is `mouseupouside` and it works great. 

Now you always should keep in mind pattern like that:
```js
obj.on("pointerdown", OnBegin);
let dragging = false;
function OnBegin(ev)
{
    const stage = //somehow find ref to stage stage
    stage.on("pointermove", OnDrag)
    stage.on("pointerup", OnEnd)
    dragging = true
}

function OnEnd()
{
    const stage = //somehow find ref to stage stage
    stage.off("pointermove", OnDrag)
    stage,off("pointerup", OnEnd)
    dragging = false
}

function OnDrag()
{
   if(!dragging) return;
   // Some usefull logics...
}

```

In case of `mousemoveoutside` and `allowOutsideMove` you don't need ref to stage which is great, and also you could't forget to remove listeners.

Yes, it looks familiar to html like apis but feels wrong in context of pixi. In html scenario we have only one stage called `document` or even `window` but with pixi it doesn't work when there is more than one pixi canvas. You still required to listen on pixi's canvas/stage because of coordinate relative to that canvas.

But still there is better way - `mousedrag` event. It sould fire on mouse move while user holds buttons. And stops when user released all buttons.
If such event exists, above snippet will be reduced to something like this:
```js
obj.on("pointerdrag", OnDrag);

OnDrag(ev)
{
   // Some usefull logics...
}
```
It just way more elegant, less error prone, and pretty common for lots of applications, to say the least.

**Comment by bigtimebuddy** [comment]
We recently added `globalpointermove`, `globaltouchmove` and `globalmousemove` as alternative events that you can subscribe to that are global (do not consider the bounds of the object being moved). These will be introduced in 7.1.1

**Reply by jakerdy** [reply]
Oh, nice, how soon it will be released?
Definnetly going to refactor all draggable items in my project!

**Reply by bigtimebuddy** [reply]
We only have one PR left to sort for 7.1.1 assuming no other major issues popup, so I'm guessing early next week?

**Comment by KreNtal** [comment]
Hello, I know it's not the correct topic but it's the only one where this is mentioned.

I'm on PIXI v7.2.2 and trying to dispatching a CustomEvent on a PIXI.Graphics but I keep getting the error:
`Error: DisplayObject cannot propagate events outside of the Federated Events API`
as @tiye mentioned.

@bigtimebuddy you said that:
> All displayObjects extend event emitter so you can emit that event

but that does not seem to work

**Reply by tiye** [reply]
my original post is outdated, also try Global Move Events in https://github.com/pixijs/pixijs/releases/tag/v7.2.0

**Reply by KreNtal** [reply]
Ok, I will do another post. Thanks for pointing me out the resource but I need CustomEvents for other purposes.

**Comment by gonulbuya** [comment]
app.stage.interactive = true;
app.stage.hitArea = app.screen;
arabaClastan2.on('pointerdown', onDragStart);
app.stage.on('pointerup', onDragEnd);
app.stage.on('pointerupoutside', onDragEnd);



function onDragStart(e) {


    this.alpha = 0.5;
    dragTarget = this;

    dragTarget.offset = {x: dragTarget.x - dragTarget.parent.toLocal(e.global).x, y: this.y - dragTarget.parent.toLocal(e.global).y};
    app.stage.on('pointermove', onDragMove);
}

function onDragMove(e) {
    
    var _this = e.currentTarget;
    
    if (dragTarget) {
        // const newPosition = _this.toLocal(event.global);
        // _this.position.x = newPosition.x+_this.offset.x;
        // _this.position.y = newPosition.y+_this.offset.y;

        //dragTarget.parent.toLocal(event.global)


         const newPosition = dragTarget.parent.toLocal(e.global)
         dragTarget.x = newPosition.x+dragTarget.offset.x;
         dragTarget.y = newPosition.y+dragTarget.offset.y;

    }

    
}

function onDragEnd(e) {

    if (dragTarget) {
        app.stage.off('pointermove', onDragMove);
        dragTarget.alpha = 1;
        dragTarget = null;
    }

    
}

**Reply by gonulbuya** [reply]
I was recently able to migrate in v7 as above.
